docs/RFC/draft-paragonie-paseto-rfc-00.txt

```diff
@@ -0,0 +1,1680 @@
+                    key = k
+                );
+
+                       Step 8: PASETO v2 decryption
+
+
+
+
+
+
+
+
+Arciszewski & Haussmann Expires October 21, 2018               [Page 13]
+
+Internet-Draft                                                April 2018
+
+
+5.3.3.  v2.Sign
+
+   Given a message "m", Ed25519 secret key "sk", and optional footer "f"
+   (which defaults to empty string):
+
+   1.  Set "h" to "v2.public."
+
+   2.  Pack "h", "m", and "f" together using PAE (see Section 2.2).
+       We'll call this "m2".
+
+   3.  Sign "m2" using Ed25519 "sk".  We'll call this "sig".  (See below
+       for pseudocode.)
+
+   4.  If "f" is:
+
+       *  Empty: return h || b64(m || sig)
+
+       *  Non-empty: return h || b64(m || sig) || "." || b64(f)
+
+       *  ...where || means "concatenate"
+
+                        sig = crypto_sign_detached(
+                            message = m2,
+                            private_key = sk
+                        );
+
+               Step 3: Generating an Ed25519 with libsodium
+
+5.3.4.  v2.Verify
+
+   Given a signed message "sm", public key "pk", and optional footer "f"
+   (which defaults to empty string):
+
+   1.  If "f" is not empty, verify that the value appended to the token
+       matches "f", using a constant-time string compare function.  If
+       it does not, throw an exception.
+
+   2.  Verify that the message begins with "v2.public.", otherwise throw
+       an exception.  This constant will be referred to as "h".
+
+   3.  Decode the payload ("sm" sans "h", "f", and the optional trailing
+       period between "m" and "f") from base64url to raw binary.  Set:
+
+       *  "s" to the rightmost 64 bytes
+
+       *  "m" to the leftmost remainder of the payload, excluding "s"
+
+
+
+
+
+Arciszewski & Haussmann Expires October 21, 2018               [Page 14]
+
+Internet-Draft                                                April 2018
+
+
+   4.  Pack "h", "m", and "f" together using PAE (see Section 2.2).
+       We'll call this "m2".
+
+   5.  Use Ed25519 to verify that the signature is valid for the
+       message: (See below for pseudocode.)
+
+   6.  If the signature is valid, return "m".  Otherwise, throw an
+       exception.
+
+                   valid = crypto_sign_verify_detached(
+                       signature = s,
+                       message = m2,
+                       public_key = pk
+                   );
+
+         Step 5: Validating the Ed25519 signature using libsodium.
+
+6.  Payload Processing
+
+   All PASETO payloads MUST be a JSON-encoded object represented as a
+   UTF-8 encoded string.  The topmost JSON object should be an object,
+   map, or associative array (depending on the language of choice), not
+   a flat array.
+
+      *Valid*:
+
+
+
+      *  "{"foo":"bar"}"
+
+      *  "{"foo":"bar","baz":12345,"678":["a","b","c"]}"
+
+      *Invalid*:
+
+
+
+      *  "[{"foo":"bar"}]"
+
+      *  "["foo"]"
+
+      *  "{0: "test"}"
```

Comment:
Isn't that just invalid JSON, as opposed to invalid use of JSON for the purposes of PASETO?